### PR TITLE
SAML idp - new sample proj encapsulates SMS OTP as an authenticator

### PR DESCRIPTION
**Pull request by AviATransmit**

### CODE REVIEW

File: saml-idp/backend/routes/index.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        headers: {
+          Authorization: `Bearer ${tokens.access_token}`,
+        },
+      });
+      const data = await resp.text();
+      console.log(`SAML response from idp: ${data}`);
+      return res.send(data);
+    }
+  } else {
+    res.send(`Login with SMS completed with error: ${req.query.error}`);
```
[the remark this answers is not in the document]
Comment (ts-ron) [reply saml-idp/backend/routes/index.js:44]: @AviATransmit this warning can be mitigated by modifying the code:
`${req.query.error}`  --> `${encodeURIComponent(req.query.error)}`

File: sample.env [hunk 2]
```diff
@@ -8,6 +8,7 @@ TS_REDIRECT_URI=<above-modified-url>/complete
 VITE_TS_CLIENT_ID=<your app client id> # VITE_ prefix allows exposing to client side pages as well as the backend
 TS_CLIENT_SECRET=<your app client secret>
 TS_APP_ID=<your app ID, will be used by the demo to point you to configuration pages>
+TS_SERVICE_PROVIDER_ID=<your testing service provider for saml idp sample>
```
Comment (ts-ron) [inline sample.env:11]: Separate this from the top variables, and add a comment? Similar to what they did with the DRS variables below
```
# SAML realted variables below
TS_SERVICE_PROVIDER_ID=<your testing service provider for saml idp sample>
```

File: saml-idp/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# SAML IDP
```
[the remark this answers is not in the document]
Comment (ts-ron) [reply saml-idp/README.md:1]: Do we have a guide?🧐

File: saml-idp/README.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+- Go to "service providers" tab and create a new provider :
+  - set "Assertion consumer service URL" : [https://samltest.id/Shibboleth.sso/SAML2/POST]
+  - set "EntityId" to [https://samltest.id/saml/sp]
+- Please save the output of SP creation which will be shown just one time
+- Bind new created SP to your application by choosing it from service providers dropdown in
+  application configuration
+
+## Build & Set new metadata
+
+- Go to [https://www.samltool.com/idp_metadata.php]
```
Comment (ts-ron) [inline saml-idp/README.md:21]: Oh no... this SAML tool is provided by a competitor ...
I suggest instead to provide a template SAML metadata file and tell the developers to embed the values in it:
```
<?xml version="1.0"?>
<md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata" validUntil="2023-05-24T13:03:12Z" cacheDuration="PT1685365392S" entityID="*****ENTITY_ID*****">
  <md:IDPSSODescriptor WantAuthnRequestsSigned="false" protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">
    <md:KeyDescriptor use="signing">
      <ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
        <ds:X509Data>
          <ds:X509Certificate>*****CERTIFICATE_CONTENT*****</ds:X509Certificate>
        </ds:X509Data>
      </ds:KeyInfo>
    </md:KeyDescriptor>
    <md:KeyDescriptor use="encryption">
      <ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
        <ds:X509Data>
          <ds:X509Certificate>*****CERTIFICATE CONTENT*****</ds:X509Certificate>
        </ds:X509Data>
      </ds:KeyInfo>
    </md:KeyDescriptor>
    <md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</md:NameIDFormat>
    <md:SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="*****HTTP_ENDPOINT*****"/>
  </md:IDPSSODescriptor>
</md:EntityDescriptor>
```

File: saml-idp/README.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+- Press "choose file" button and locate your metadata file you created before and press "UPLOAD" (
+  you need to see a confirmation with the xml metadata file after few seconds)
+
+### Sample user experience
+
+- Extract service provider id from the "SAML 2.0 endpoint (HTTP)" param (which you have saved from
+  the creation of the SP in transmit portal) example of "SAML 2.0 endpoint (HTTP)" output :
+  [https://api.userid.security/serviceprovider/saml2/sso/fn3jm0zqh0xfvvn38k327]
+  ("fn3jm0zqh0xfvvn38k327" is the SP id)
+- create .env file in "saml-idp" folder and add : TS_SERVICE_PROVIDER_ID={TS_SERVICE_PROVIDER_ID} (
```
[the remark this answers is not in the document]
Comment (AviATransmit) [reply saml-idp/README.md:38]: yes .. your'e right 
i got @Yair's attention for those places which he should fix

File: sample.env (hunk 2 above)
Comment (AviATransmit) [reply sample.env:13]: done

File: saml-idp/README.md (hunk 4 above)
Comment (AviATransmit) [reply saml-idp/README.md:21]: oh ... yes, absolutely

File: saml-idp/README.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+## Build & Set new metadata
+
+- Go to [https://www.samltool.com/idp_metadata.php]
+- Set "EntityId" to be the "identity provider issuer" & "SP X.509 cert" to be the "x.509
+  certificate" (values saved from SP creation)
+- Set "Single Sign On Service Endpoint" to be directed to "sample login app"
+  [http://localhost:8080/saml/authn]
+- Press the "BUILD IDP METADATA" and save xml data into a file
+- Go to [https://samltest.id/start-idp-test/] and press the link at first line ["metadata link
```
[the remark this answers is not in the document]
Comment (AviATransmit) [reply saml-idp/README.md:27]: sure

File: saml-idp/backend/routes/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (AviATransmit) [reply saml-idp/backend/routes/index.js:45]: done

File: saml-idp/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (AviATransmit) [reply saml-idp/README.md:1]: so we said it won't including a dedicated guide ( just a ref to "https://developer.transmitsecurity.com/guides/user/sso_login_saml/" )

File: saml-idp/README.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+- let's choose a SAML IDP tester simulating SP ( for our example let's open
+  [https://samltest.id/start-idp-test/])
+
+## Set up the Transmit application & service provider
+
+- Go to your tenant pages on the [Transmit portal](https://portal.identity.security/), and create
+  new application
+- Go to "service providers" tab and create a new provider :
+  - set "Assertion consumer service URL" : [https://samltest.id/Shibboleth.sso/SAML2/POST]
+  - set "EntityId" to [https://samltest.id/saml/sp]
```
[the remark this answers is not in the document]
Comment (AviATransmit) [reply saml-idp/README.md]: add it to .README

File: saml-idp/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (AviATransmit) [reply saml-idp/README.md]: done

**Review by ts-ron — APPROVED** [review]
LGTM, great work @AviATransmit

**Pull request merged**